### ISSUE TITLE
Wrong output for a simple function that returns a constant - happen on ZLinux

### ISSUE THREAD

**Issue by tungld** [onnx/onnx-mlir#816]
While checking the PR #815 I found that a simple program failed on ZLinux:

A test program that just prints out a constant tensor:
```mlir
module  {
  func @main_graph() -> tensor<2xf32> {
    %0 = "onnx.Constant"() {value = dense<[3.000000e+00, 2.000000e+00]> : tensor<2xf32>} : () -> tensor<2xf32>
    return %0 : tensor<2xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph, numInputs = 0 : i32, numOutputs = 1 : i32, signature = "[ \0A]\00@[   { \22type\22 : \22f32\22 , \22dims\22 :
 [2] , \22name\22 : \22z\22 }\0A\0A]\00"} : () -> ()
}
```
Compile the code using `onnx-mlir` and run using Python or C, output is `[2., 3.]` (it should be `[3., 2.]`).

Checking the constant in the generated LLVM IR, it is correctly generated:
```
llvm.mlir.global internal constant @constant_0(dense<[3.000000e+00, 2.000000e+00]> : tensor<2xf32>) : !llvm.array<2 x f32>
```

However, using `-O0` instead of `-O3` in the llvm `opt` command produces the correct result.

**Comment by tungld** [comment]
Now, we generate LLVMIR using `onnx-mlir-opt` but run the LLVMIR code using `mlir-cpu-runner` with `-O0` or `-O3`.

Since `mlir-cpu-runner` does not support tensors as result, we change our code so that only the first element in the constant tensor is returned.

```mlir
module  {
  func @main_graph() -> f32 {
    %cst0 = constant 0 : index
    %0 = "onnx.Constant"() {value = dense<[3.000000e+00, 2.000000e+00]> : tensor<2xf32>} : () -> memref<2xf32>
    %1 = krnl.load %0[%cst0] : memref<2xf32>
    return %1 : f32
  }
}
```

Compile and run:
```bash
$ onnx-mlir-opt --convert-onnx-to-krnl constant.mlir  --convert-krnl-to-affine --convert-krnl-to-llvm | ~/dl/llvm-project/build/bin/mlir-cpu-runner -O3 -e main_graph --entry-point-result=f32
$ 3.000000e+00
```
We got the correct result here for both `-O0` and `-O3` in `opt`.

Some potential places in `onnx-mlir` that cause the issue:
- `onnx.EntryPoint`
- Tools that are used between the LLVMIR level and the binary output.

**Comment by tungld** [comment]
This is definitely a bug in llvm opt, which does constant folding to replace memcpy with load/store **when memcpy length is 1/2/4/8**: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp#L135
```
// If MemCpyInst length is 1/2/4/8 bytes then replace memcpy with
// load/store.
  ConstantInt *MemOpLength = dyn_cast<ConstantInt>(MI->getLength());
  if (!MemOpLength) return nullptr;
```


In our example above, memcpy is generated by MLIR to copy data from the constant to the returned Memref.  The memcpy length is 8.

Let's use a constant tensor with size > 8, like this:
```mlir
module  {
  func @main_graph() -> tensor<3xf32> {
    %0 = "onnx.Constant"() {value = dense<[113.000000e+00, 311.000000e+00, 131.000000e+00]> : tensor<3xf32>} : () -> tensor<3xf32>
    return %0 : tensor<3xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph, numInputs = 0 : i32, numOutputs = 1 : i32, signature = "[ \0A]\00@[   { \22type\22 : \22f32\22 , \22dims\22
: [3] , \22name\22 : \22z\22 }\0A\0A]\00"} : () -> ()
}
```
In this case, `onnx-mlir` produced the correct result `[113., 311., 131.]`.

**Comment by AlexandreEichenberger** [comment]
@tungld thanks for your careful evaluation. If the bug is only on Z, should we forward it to the German support team? Or does it occur on all platforms?

**Comment by tungld** [comment]
Yes, it is only on Z. I'll try to see if it is easy to fix or not. If it is non-trivial, I will forward it to the German support team.

**Comment by tungld** [comment]
Found the root cause of this issue, and possibly of all constants related issues we have got on Z so far.

It was because LLVM opt uses module's DataLayout to know whether data is in little or big endian, e.g. `bool isLittleEndian = DL.isLittleEndian();` where DL is the module's DataLayout.
If there is no DataLayout information, it is little endian by default. So all optimizations related to constants suppose that data is in little endian, which causes issues on Z.

A proper fix, I think, is to annotate ModuleOp with endian information so that LLVM optimizations correctly know the endianness of the data. I created a PR #819 in our onnx-mlir for such an annotation.
